### PR TITLE
Implemented ColumnCapper and its tests

### PR DESCRIPTION
**Pull request by arthurpaulino**
Issue: https://github.com/koaning/scikit-lego/issues/66

### CODE REVIEW

**Comment by arthurpaulino** [comment]
Oops, forgot to check flake8.

**Comment by arthurpaulino** [comment]
Alright, this one was tricky: https://github.com/koaning/scikit-lego/issues/120

File: sklego/preprocessing.py [hunk 1]
```diff
@@ -206,3 +206,179 @@ def _check_X_for_type(X):
...
+    2  3   inf      3.00      13.6
+    3  4  14.0      3.70      13.6
+    >>> capper = ColumnCapper(discard_infs=True) # Discarding infs
+    >>> df[['a', 'b']] = capper.fit_transform(df[['a', 'b']]) # Transforming multiple columns
+    >>> df
+          a     b  a_capped  b_capped
+    0  1.15  11.1      1.15      11.1
+    1  2.00  12.0      2.00      12.0
+    2  3.00   NaN      3.00      13.6
+    3  3.85  13.8      3.70      13.6
```
Comment (koaning) [inline sklego/preprocessing.py]: i wonder ... why are we not seeing `13.8` at `b_capped` here?

File: tests/test_preprocessing/test_columncapper.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    estimator_checks.check_methods_subset_invariance,
+    estimator_checks.check_transformer_data_not_an_array,
+    estimator_checks.check_transformer_general,
+]))
+def test_estimator_checks(test_fn):
+    test_fn(ColumnCapper.__name__, ColumnCapper())
+
+
+def test_quantiles():
+    def expect_type_error(min_quantile=0.05, max_quantile=0.95):
```
Comment (koaning) [inline tests/test_preprocessing/test_columncapper.py]: never seen this pattern before but i like it.

File: sklego/preprocessing.py (hunk 1 above)
Comment (koaning) [reply sklego/preprocessing.py]: I also notice that `np.inf` values are filtered out during fit but `NaN` values are not. Is there a reason for this?

File: tests/test_preprocessing/test_columncapper.py (hunk 2 above)
Comment (arthurpaulino) [reply tests/test_preprocessing/test_columncapper.py]: I was improvising. It was my first time implementing automated tests :B

File: sklego/preprocessing.py (hunk 1 above)
Comment (arthurpaulino) [reply sklego/preprocessing.py]: Hello, @koaning 

### On the examples

> i wonder ... why are we not seeing `13.8` at `b_capped` here?

I used different quantiles for the second capping... maybe it was not an intuitive move. Do you think I should use the same capping quantiles or keep them different and point out to the fact that I changed the capping values?

### On *inf* cells

> I also notice that `np.inf` values are filtered out during fit but `NaN` values are not. Is there a reason for this?

Yes. The *inf* part came from [here](https://github.com/koaning/scikit-lego/issues/66#issuecomment-486374288) (typo fixed):

> This ColumnCapper could then take all `inf` values and cast them to the 95th percentile and `-inf` to the 5th.

And from [here](https://github.com/koaning/scikit-lego/issues/66#issue-424303629):

> In ML some time your features have extreme large values or even infinite value (np.inf), we want to cap those values with a feature transformer.

Leaving *inf* cells untouched doesn't make much sense to me since we're attempting to guarantee capping limits. But this is just my point of view...

As an attempt to contemplate what I said [here](https://github.com/koaning/scikit-lego/issues/66#issuecomment-486225626):

> I usually replace them by np.nan and then deal with those from there.

I made it an option with the `discard_infs` parameter and pointed out the importance of using it depending on how those *inf* cells were produced. But its default value is `False`, which makes `ColumnsCapper` simply cap them.

### On *nan* cells

Unfortunately, we can't know what to do with the *nan* cells, as you said [here](https://github.com/koaning/scikit-lego/issues/66#issuecomment-486374288):

> `Na` remains tricky, but thats what valueimputers are for.

So I left those out as a responsibility for the user. Enforcing some artificial value on them with this transformer doesn't make much sense to me. But again, this is just my point of view. What do you think?

Comment (arthurpaulino) [reply sklego/preprocessing.py]: Hello @koaning!

> i wonder ... why are we not seeing `13.8` at `b_capped` here?

I used different quantiles for the second capping... maybe it was not an intuitive move. Do you think I should use the same capping quantiles or keep them different and point out to the fact that I changed the capping values?

> I also notice that np.inf values are filtered out during fit but NaN values are not. Is there a reason for this?

Yeah, `pandas` ignores *nan* cells when computing the quantiles. But if I leave *inf* cells there, they will pull the quantiles towards them.

```python
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame([1, 2, 3, 4, 5, np.inf])
>>> df.quantile([0.05, 0.95])
         0
0.05  1.25
0.95   inf
```

File: tests/test_preprocessing/test_columncapper.py (hunk 2 above)
Comment (MBrouns) [reply tests/test_preprocessing/test_columncapper.py]: good improv, I like it!

File: sklego/preprocessing.py (hunk 1 above)
Comment (MBrouns) [reply sklego/preprocessing.py]: That makes sense

**Comment by arthurpaulino** [comment]
@MBrouns, thanks for the detailed reviews. Your ideas were very insightful and I'll make the changes soon.

Oh wow, what a learning experience! :smile:

File: tests/test_preprocessing/test_columncapper.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+from sklego.common import flatten
+from sklego.preprocessing import ColumnCapper
+from tests.conftest import nonmeta_checks
+
+# ColumnCapper works with nan/inf cells
+nonmeta_checks_allow_nan_inf = [test for test in nonmeta_checks if test.__name__ != 'check_estimators_nan_inf']
+
+
+@pytest.mark.parametrize("test_fn", flatten([
```
[the remark this answers is not in the document]
Comment (arthurpaulino) [reply tests/test_preprocessing/test_columncapper.py]: Wouldn't failing at some test block the PR? Or are you saying that you would manually merge it even with failed tests?

[the remark this answers is not in the document]
Comment (MBrouns) [reply tests/test_preprocessing/test_columncapper.py]: Ah yes, I should've been more clear. If possible, it's best to use the lists containing all tests for a specific category of transformer. 

If that isn't possible because on of the tests in that list can't pass, copy the contents of the list to the test parameterization and put a comment in the body to indicate which test you removed and why.

**Comment by arthurpaulino** [comment]
Another detail to consider... I think it would be nice to use the [sklearn.preprocessing.RobustScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.RobustScaler.html) standard to provide quantiles:

```python
RobustScaler(with_centering=True, with_scaling=True,
             quantile_range=(25.0, 75.0), copy=True)
```

They use a tuple instead of two separate parameters.

**Comment by koaning** [comment]
might it be nice to allow the user to select the interpolation method used in pandas? looking at the quantile method is seems like there are subtle differences.

**Comment by koaning** [comment]
One final comment from my side at the moment: it might be nice to add a snippet to the documentation. The examples from the docstring are automatically generated but the adoption rate might be even better if there's a separate article attached to our current [preprocessors docs](https://scikit-lego.readthedocs.io/en/latest/preprocessing.html).

**Comment by arthurpaulino** [comment]
> might it be nice to allow the user to select the interpolation method used in pandas? looking at the quantile method is seems like there are subtle differences.
> 
> maybe add a `method=` param?

Great idea. And it's also available on the [numpy API](https://docs.scipy.org/doc/numpy/reference/generated/numpy.quantile.html), which I am planning to use to avoid those conversions to `pandas.DataFrame`, as @MBrouns said.

But isn't `method` too generic? For those who are familiar with pandas/numpy APIs, `interpolation` would be more likely to ring a bell.

> One final comment from my side at the moment: it might be nice to add a snippet to the documentation. The examples from the docstring are automatically generated but the adoption rate might be even better if there's a separate article attached to our current [preprocessors docs](https://scikit-lego.readthedocs.io/en/latest/preprocessing.html).

Good idea. I am a fan of more precise documentation like [sklearn's User Guide](https://scikit-learn.org/stable/user_guide.html).

**Comment by koaning** [comment]
@arthurpaulino agreed -> interpolation > method

**Comment by arthurpaulino** [comment]
@koaning I'm afraid I won't have enough time to write a more in-depth explanation of `ColumnCapper` on the [preprocessors docs](https://scikit-lego.readthedocs.io/en/latest/preprocessing.html) right now but I think it would be nice to have an explicit explanation of each interpolation method in the future.

**Comment by MBrouns** [comment]
Good to hear the min-max thing worked! Was the problem that you want to ignore the NaN values but the numpy method doesn't? if so, there is a [`nanquantile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanquantile.html#numpy.nanquantile) function. Otherwise keeping it in pandas is also fine.

**Comment by koaning** [comment]
I'll update to master branch and merge! 

I also wish to take a moment to point out that this is amazing ... 

... a mere two months ago I was discussing with @MBrouns that it might be a good idea to have an open source library as part of our training curriculum. It seemed awesome to teach people how to contribute but we also noticed that colleagues were writing the same transformers/estimators over and over again at clients. 

The reception of the package has been good and there are companies in the Netherlands that have shown interest in contributing. Then we noticed we saw a pull request from somebody who lives on the other side of the planet who we never met. 

What a time to be alive. 

Obrigado!

**Comment by arthurpaulino** [comment]
@koaning @MBrouns 
Thanks for the opportunity. I am learning a LOT!

**Comment by arthurpaulino** [comment]
@MBrouns Another thing that was holding me back was the *inf* filter. I thought `pandas.DataFrames` did something smarter behind the scenes but my tests confirmed otherwise. Let me show you.

```python
>>> arr = np.array([[1, 2], [3, np.inf]])
>>> arr
array([[ 1.,  2.],
       [ 3., inf]])
>>> arr[arr<np.inf]
array([1., 2., 3.])
```

The mask flattens the `numpy.array`, which is undesirable. It does not happen with `pandas.DataFrame`:

```python
>>> df = pd.DataFrame(arr)
>>> df
     0    1
0  1.0  2.0
1  3.0  inf
>>> df[df<np.inf]
     0    1
0  1.0  2.0
1  3.0  NaN
```

Using `np.nanquantile` works there just like using the `quantile` method. But would it be possible to create an efficient *inf* filter? I thought `pandas` did not copy the array to perform the *inf* filter, but I was wrong:

```python
>>> df_ = df[df<np.inf]
>>> df_
     0    1
0  1.0  2.0
1  3.0  NaN
>>> arr[0,0] = 10  # this will change df. will it change df_, though?
>>> df
      0    1
0  10.0  2.0
1   3.0  inf
>>> df_
     0    1
0  1.0  2.0
1  3.0  NaN
```

It means that, instead of instantiating a `pandas.DataFrame` object, it would be possible to simply make a manual copy of the `numpy.array` and use the `np.putmask` to filter its *inf* cells. And then use the `np.nanquantile` to compute the quantiles even with *nan* cells. Copying the data is something that `pandas` already does, as shown above. The advantage of this approach is to avoid the conversion overhead.

I will make a PR soon to optimize the `fit` method without the need for `pandas.DataFrame` methods :)

**Comment by koaning** [comment]
@arthurpaulino ill have a look at you merge request after pydata amsterdam. both @MBrouns and myself are in the committee. I did however, push your changes to the current release (0.1.6). so congrats, your changes are live :)
